File: skills/vector-index/scripts/references.py
#!/usr/bin/env python3
"""
references — extract, validate, and ground external references / citations.

Generalized from the company-data MCP server and citation engine (C8): pull the
references out of a piece of text (URLs and citation-shaped tokens), and validate
each one — either against the indexed corpus (does this citation actually appear
in what we know?) or, optionally and off the hot path, against the live network.
Unverifiable references are flagged `[UNVERIFIED]`, the same defense the legal-RAG
document prescribes for hallucinated citations.

Stdlib only. The corpus check takes an injected `search_fn`, so it is decoupled
from the engine and unit-testable with a fake searcher; network resolution is
opt-in.
"""

from __future__ import annotations

import re
import urllib.request

_URL_RE = re.compile(r"https?://[^\s)<>\]\"']+")
# Generic citation shapes: "228/1929", "KPL 3:9.2", "RFC 7231", "ISO 8601",
# "Section 12", "[Law §]". Domain callers can pass extra patterns.
_CITATION_RES = [
    re.compile(r"\b[0-9]{1,5}/[0-9]{2,4}\b"),                  # number/year
    re.compile(r"\b[A-Z]{2,6}\s?\d+(?::\d+(?:\.\d+)*)?\b"),    # CODE 3:9.2 / RFC 7231
    re.compile(r"§\s?\d+[a-z]?"),                              # § 36
]
# Real URLs/citations are short. An over-long match is almost always a minified
# line or an embedded blob — and would overflow the unified-DB's btree
# UNIQUE(kind, uri) index (~2704-byte limit), killing the extract_references job.
_MAX_URI = 2048


def extract_references(text: str, patterns: list[re.Pattern] | None = None) -> list[dict]:
    """Return de-duplicated [{kind, uri}] for URLs and citation-shaped tokens."""
    out: list[dict] = []
    seen: set[tuple[str, str]] = set()

    def _add(kind: str, uri: str):
        if not uri or len(uri) > _MAX_URI:
            return
        key = (kind, uri)
        if key not in seen:
            seen.add(key)
            out.append({"kind": kind, "uri": uri})

    for u in _URL_RE.findall(text or ""):
        _add("url", u.rstrip(".,);"))
    for pat in (patterns or _CITATION_RES):
        for m in pat.findall(text or ""):
            _add("citation", m.strip())
    return out


def validate_citations(
    text: str,
    search_fn,
    *,
    threshold: float = 0.0,
    topk: int = 5,
    patterns: list[re.Pattern] | None = None,
) -> dict:
    """Check every reference in `text` against the corpus via `search_fn`.

    `search_fn(query, topk)` must return a result dict shaped like the engine's
    (a "results" list of {text, source, ...}). A reference is *verified* when it
    appears verbatim in a retrieved chunk, or a retrieved chunk clears `threshold`
    on rerank/vector score. Returns the per-reference verdicts plus an annotated
    copy of the text with unverifiable references marked `[UNVERIFIED]`.
    """
    refs = extract_references(text, patterns)
    annotated = text
    verdicts = []
    for ref in refs:
        uri = ref["uri"]
        res = search_fn(uri, topk) or {}
        results = res.get("results", [])
        verbatim = next((r for r in results if uri.lower() in (r.get("text", "").lower())), None)
        if verbatim is not None:
            ok, src = True, verbatim.get("source") or verbatim.get("url")
        elif results:
            top = results[0]
            score = top.get("rerank_score", top.get("vector_score") or 0.0)
            ok = score is not None and score >= threshold
            src = top.get("source") or top.get("url") if ok else None
        else:
            ok, src = False, None
        verdicts.append({"kind": ref["kind"], "uri": uri, "verified": ok, "source": src})
        if not ok:
            annotated = annotated.replace(uri, f"{uri} [UNVERIFIED]")
    n_ok = sum(1 for v in verdicts if v["verified"])
    return {
        "references": verdicts,
        "verified": n_ok,
        "unverified": len(verdicts) - n_ok,
        "annotated": annotated,
    }


def resolve_reference(uri: str, *, timeout: float = 8.0, network: bool = False) -> dict:
    """Optionally validate a URL is live (HEAD request). Off by default to honor
    the plugin's no-network-at-query-time guarantee; callers opt in explicitly."""
    if not network or not uri.lower().startswith(("http://", "https://")):
        return {"uri": uri, "checked": False, "reachable": None}
    req = urllib.request.Request(uri, method="HEAD", headers={"User-Agent": "vectors-plugin"})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return {"uri": uri, "checked": True, "reachable": True, "status": resp.status}
    except Exception as e:  # network/HTTP error -> not reachable
        return {"uri": uri, "checked": True, "reachable": False, "error": str(e)[:200]}

File: src/search/references.ts
/**
 * references — extract, validate, and ground external references / citations.
 *
 * Pull references (URLs and citation-shaped tokens) out of text, then validate
 * each — against the indexed corpus (does this citation actually appear in what
 * we know?) or, opt-in and off the hot path, against the live network.
 * Unverifiable references are flagged `[UNVERIFIED]`, the defense against
 * hallucinated citations.
 *
 * The corpus check takes an injected `searchFn`, decoupling it from the engine
 * (unit-testable with a fake searcher); network resolution is opt-in.
 */

const URL_RE = /https?:\/\/[^\s)<>\]"']+/g

/**
 * Generic citation shapes: "228/1929", "KPL 3:9.2", "RFC 7231", "ISO 8601",
 * "SECTION 36". Domain callers can pass extra patterns.
 */
const CITATION_RES: RegExp[] = [
  /\b\d{1,5}\/\d{2,4}\b/g, // number/year
  /\b[A-Z]{2,6}\s?\d+(?::\d+(?:\.\d+)*)?\b/g, // CODE 3:9.2 / RFC 7231
  /§\s?\d+[a-z]?/g, // section-sign 36
]

/**
 * Real URLs/citations are short. An over-long match is almost always a minified
 * line or embedded blob — and would overflow the unified-DB's btree
 * UNIQUE(kind, uri) index (~2704-byte limit), killing the extract job.
 */
const MAX_URI = 2048

export type RefKind = 'url' | 'citation'
export interface Reference {
  kind: RefKind;
  uri:  string;
}

/** A corpus result row as the engine returns it (shape-compatible subset). */
export interface SearchResultRow {
  text?:         string;
  source?:       string | null;
  url?:          string | null;
  rerank_score?: number | null;
  vector_score?: number | null;
}
export interface SearchFnResult {
  results?: SearchResultRow[];
}
export type SearchFn = (
  query: string,
  topk: number,
) => SearchFnResult | null | undefined

/** Ensure a regex is global so `matchAll` yields every occurrence. */
function asGlobal (re: RegExp): RegExp {
  return re.global ? re : new RegExp(re.source, re.flags + 'g')
}

/** Return de-duplicated [{kind, uri}] for URLs and citation-shaped tokens. */
export function extractReferences (
  text: string,
  patterns?: RegExp[],
): Reference[] {
  const out: Reference[] = []
  const seen             = new Set<string>()

  const add = (kind: RefKind, uri: string) => {
    if (!uri || uri.length > MAX_URI)
      return

    const key = kind + ' ' + uri
    if (!seen.has(key)) {
      seen.add(key)
      out.push({ kind, uri })
    }
  }

  const src = text || ''
  for (const m of src.matchAll(asGlobal(URL_RE)))
    add('url', m[0].replace(/[.,);]+$/, ''))
  for (const pat of patterns ?? CITATION_RES)
    for (const m of src.matchAll(asGlobal(pat)))
      add('citation', m[0].trim())
  return out
}

export interface CitationVerdict {
  kind:     RefKind;
  uri:      string;
  verified: boolean;
  source:   string | null;
}
export interface CitationReport {
  references: CitationVerdict[];
  verified:   number;
  unverified: number;
  annotated:  string;
}

/**
 * Check every reference in `text` against the corpus via `searchFn`. A reference
 * is verified when it appears verbatim in a retrieved chunk, or a retrieved
 * chunk clears `threshold` on rerank/vector score. Returns per-reference
 * verdicts plus a copy of the text with unverifiable refs marked `[UNVERIFIED]`.
 */
// eslint-disable-next-line complexity -- citation verification fans over several ref shapes
export function validateCitations (
  text: string,
  searchFn: SearchFn,
  opts: { threshold?: number; topk?: number; patterns?: RegExp[] } = {},
): CitationReport {
  const { threshold = 0.0, topk = 5, patterns } = opts
  const refs                                    = extractReferences(text, patterns)
  let annotated = text
  const verdicts: CitationVerdict[] = []

  for (const ref of refs) {
    const uri      = ref.uri
    const res      = searchFn(uri, topk) || {}
    const results  = res.results ?? []
    const verbatim = results.find(r =>
      (r.text ?? '').toLowerCase().includes(uri.toLowerCase()),
    )
    let ok: boolean
    let source: string | null
    if (verbatim) {
      ok = true
      source = verbatim.source ?? verbatim.url ?? null
    }
    else if (results.length > 0) {
      const top   = results[0]
      const score = top.rerank_score ?? top.vector_score ?? 0.0
      ok = score != null && score >= threshold
      source = ok ? top.source ?? top.url ?? null : null
    }
    else {
      ok = false
      source = null
    }
    verdicts.push({ kind: ref.kind, uri, verified: ok, source })
    if (!ok)
      annotated = annotated.split(uri).join(uri + ' [UNVERIFIED]')
  }

  const nOk = verdicts.filter(v => v.verified).length
  return {
    references: verdicts,
    verified:   nOk,
    unverified: verdicts.length - nOk,
    annotated,
  }
}

export interface ResolveResult {
  uri:       string;
  checked:   boolean;
  reachable: boolean | null;
  status?:   number;
  error?:    string;
}

/**
 * Optionally validate a URL is live (HEAD request). Off by default to honor the
 * plugin's no-network-at-query-time guarantee; callers opt in explicitly.
 */
export async function resolveReference (
  uri: string,
  opts: { timeout?: number; network?: boolean } = {},
): Promise<ResolveResult> {
  const { timeout = 8.0, network = false } = opts
  const lower                              = uri.toLowerCase()
  if (
    !network ||
    !(lower.startsWith('http://') || lower.startsWith('https://'))
  )
    return { uri, checked: false, reachable: null }

  const ctrl  = new AbortController()
  const timer = setTimeout(() => ctrl.abort(), timeout * 1000)
  try {
    const resp = await fetch(uri, {
      method:  'HEAD',
      headers: { 'User-Agent': 'vectors-plugin' },
      signal:  ctrl.signal,
    })
    return { uri, checked: true, reachable: true, status: resp.status }
  }
  catch (e) {
    return {
      uri,
      checked:   true,
      reachable: false,
      error:     String(e instanceof Error ? e.message : e).slice(0, 200),
    }
  }
  finally {
    clearTimeout(timer)
  }
}

/**
 * Groundedness convenience: given a `corpus` of strings, report which references
 * in `text` appear verbatim within any corpus entry. Wraps `validateCitations`
 * with a trivial in-memory searcher.
 */
export function checkGroundedness (
  text: string,
  corpus: string[],
): CitationReport {
  const searchFn: SearchFn = query => ({
    results: corpus
      .filter(c => c.toLowerCase().includes(query.toLowerCase()))
      .map(c => ({ text: c, source: null })),
  })
  return validateCitations(text, searchFn)
}
